TINKERPOP-2224 Detect and fix iterator leaks

divijvaidya opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-2224

## Motivation
During the execution of the query the underlying provider layer might open some resources such as cursors or iterators which should be closed on completion of execution of the query. This is largely done by calling Traversal.close() method at the end of the traversal.
However, there are bugs in the query processor which lead to a leak of these resources. 

As an example, consider the following query:

`g.V().not(__.V().out()).fold().unfold()`

The implementation of not() step creates a child traversal which would open an iterator(aka resource) at the storage layer but due to a bug in TraversalUtil.java, it will never close this traversal (even on calling Traversal.close() on the global Traversal). This leads to open resources on the storage layer.

The purpose of this task is to add a framework for detecting an iterator leak and consequently fixing the bugs unearthed by the new framework.

## Changes
1. Add a framework to the existing test suite which would detect iterator leak during execution of every query in the test suite.
2. Fix the bugs causing iterator leak at multiple places
3. Provide an annotation to skip check for iterator leak which require fixing. To be taken up as a separate task.

## Testing
- After adding the new assertion, the existing test cases fail and after fixing the bugs, they pass.
- gremlin-core: mvn clean install -DskipIntegrationTests=fals
- gremlin-test: mvn clean install -DskipIntegrationTests=fals
- tinkergraph-gremlin: mvn clean install -DskipIntegrationTests=fals

divijvaidya commented:
The failing travis check seems to be unrelated to this code change:

```
[ERROR] COMPILATION ERROR : 
[ERROR] /home/travis/build/apache/tinkerpop/spark-gremlin/src/test/java/org/apache/tinkerpop/gremlin/spark/structure/io/SparkContextStorageCheck.java:[48,14] cannot find symbol
  symbol: method tearDown()
[ERROR] /home/travis/build/apache/tinkerpop/spark-gremlin/src/test/java/org/apache/tinkerpop/gremlin/spark/structure/io/SparkIoRegistryCheck.java:[48,14] cannot find symbol
  symbol: method tearDown()
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:testCompile (default-testCompile) on project spark-gremlin: Compilation failure: Compilation failure:
[ERROR] /home/travis/build/apache/tinkerpop/spark-gremlin/src/test/java/org/apache/tinkerpop/gremlin/spark/structure/io/SparkContextStorageCheck.java:[48,14] cannot find symbol
[ERROR] symbol: method tearDown()
[ERROR] /home/travis/build/apache/tinkerpop/spark-gremlin/src/test/java/org/apache/tinkerpop/gremlin/spark/structure/io/SparkIoRegistryCheck.java:[48,14] cannot find symbol
```

spmallette commented:
I haven't looked into this PR in detail yet given code freeze and release week, but note that all release branches including `master` are building fine on travis. I've re-started Travis jobs for this PR twice now I think and they keep failing the same way. You may wish to rebase to see if that fixes things. Further note that `master` is now 3.5.0.

divijvaidya commented:
I found the reason for travis failures. My change in `AbstractGremlinTest` caused it (my bad for not spotting it earlier). I have fixed the issue and hopefully, they should all pass now.

The current code is also rebased with the latest master.

spmallette commented:
Looks green on travis now. Ultimately I'd think you would want to rebase on `tp34` (though that branch is not back on SNAPSHOT yet) as I assume you'd like to see this fix go into the 3.4.x line as 3.5.0 really doesnt' have a release timeline yet.

spmallette commented:
In reviewing the code of this in more detail, this looks like a good change to me. I think it would be good to see it rebased on `tp33` so that we get this fix on all current release lines.  

The "leak detector" is neat and a reference implementation in TinkerGraph seems like a fine place for it since it cores our testing.  I see that you had to add an annotation to the test framework to "ignore leaks" which might be a helpful tool. Could you comment on the tests that you've assigned the annotation currently? Were those leaks hard to locate and fix?

divijvaidya commented:
This PR is rebased from tp33 now.

> "Could you comment on the tests that you've assigned the annotation currently? Were those leaks hard to locate and fix?"

I did not get time to look at those tests yet. I will take them up in the coming weeks.

spmallette commented on tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/process/traversal/step/sideEffect/TinkerGraphStep.java:
```diff
@@ -27,26 +34,24 @@
 import org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerHelper;
 import org.apache.tinkerpop.gremlin.util.iterator.IteratorUtils;
 
-import java.util.ArrayList;
```
I see a number of changes where the imports are reordered in this fashion. that ordering here is out of step with our style. looking at other classes, we tend to keep the list the `java.*` ones after other libraries. i guess i mostly call your attention to it in case your IDE is automatically making these changes you can adjust the settings so that future PRs don't have these sorts of changes inadvertently mixed in, but i also think it would be good to reverse these sorts of changes in this PR to avoid unnecessary modifications.

divijvaidya replied:
Thanks. I was about to ask this question in dev email list. We should probably have an intelliJ/eclipse configuration which the developers can import in the IDE and not have to worry about code styling issues.

spmallette commented:
Note that we've re-opened the releases branches for development so `tp33` is now on 3.8.0-SNAPSHOT. 

> I did not get time to look at those tests yet. I will take them up in the coming weeks.

ok - i'm sorry, but it's not clear from your response if you intend to make those changes on this PR or a future one. i think i'm ok to merge this as-is because it's not really introducing a new problem, so much as shedding light on one. i just wouldn't want to forget that we have those leaks that need review.

upon further thought, i think it would be nice if this PR had some Upgrade Documentation for providers to call their attention to what's been done here.  Maybe there should even be some details in the [Provider documentation](http://tinkerpop.apache.org/docs/current/dev/provider/#_implementing_gremlin_core)? Including some docs of this nature would be a nice addition, but I can probably write something up on merge if you like. Just let me know.

divijvaidya commented:
1. Added provider documentation, please feel free to modify it as you deem fit.
2. Started a JIRA to track the task of fixing the remaining JIRA's. https://issues.apache.org/jira/browse/TINKERPOP-2238
3. Rebased with tp33.

This PR is now ready for review.

spmallette commented on docs/src/dev/provider/index.asciidoc:
```diff
@@ -805,6 +805,16 @@ environment variable, depending on your project layout. Some tests require this
 SureFire Plugin, this is done via the configuration argLine with `-Dbuild.dir=${project.build.directory}`.
 
+==== Checking resource leak
+Tinkerpop query engine retrieves data by interfacing with the provider using iterators. These iterators (depending
```
nit: "Tinkerpop" - the inner "p" is capitalized as in "TinkerPop"

spmallette commented:
Looks like there might be a problem somewhere - travis is complaining of:

```text
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/apache/tinkerpop/gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/WhereTest.java:[276,146] method convertToVertex in class org.apache.tinkerpop.gremlin.AbstractGremlinTest cannot be applied to given types;
  required: org.apache.tinkerpop.gremlin.structure.Graph,java.lang.String
  found: java.lang.String
  reason: actual and formal argument lists differ in length
```

divijvaidya commented:
Strangely, I am not able to repro the latest test failure on my local setup (mvn clean install -DskipIntegrationTests=false on tinkergraph is successful). I am looking into this. Will update the PR when I have fixed it.

spmallette commented:
i just restarted the job - i wonder if there is something non-deterministic in there somehow....

divijvaidya commented:
Disabled leak check on failing tests. This is ready for review now.

spmallette commented on gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/AbstractGremlinTest.java:
```diff
@@ -173,7 +189,7 @@ public Object convertToVertexId(final Graph graph, final String vertexName) {
     public Vertex convertToVertex(final Graph graph, final String vertexName) {
         // all test graphs have "name" as a unique id which makes it easy to hardcode this...works for now
-        return graph.traversal().V().has("name", vertexName).next();
+        return graphProvider.traversal(graph).V().has("name", vertexName).toList().get(0);
```
Did you learn something here? Is there a need for the test suite to prefer `graphProvider.traversal(graph)` over simply `graph.traversal()`? I just see that you made such changes here, but in other places you didn't make such changes:

https://github.com/apache/tinkerpop/pull/1118/files#diff-d201caccfbd94491ed409d1f26ec5b62R149

divijvaidya replied:
The purpose of this change was to consume the iterator completely so that the underlying resources are closed and hence, the `toList().get(0)` is the significant part in the code change above. The traversal can be obtained in any manner. I have reverted the first part to use graph.traveral() itself to minimize changes.

spmallette replied:
i just started to merge this and i think that change might have been an artifact from your rebase at some point as i see that style in `tp34`. sorry about making you change that.  i will probably go back and add that change to `tp33` after this merge is done.

spmallette commented:
VOTE +1
